common/migration/default-comparison.go

```diff
@@ -0,0 +1,157 @@
+// The MIT License (MIT)
+
+// Copyright (c) 2017-2020 Uber Technologies Inc.
+
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+package migration
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+)
+
+var _ ComparisonFn[struct{}] = defaultComparisonFn[struct{}]
+
+// DefaultComparisonFn is a simple, suitable-for-low-throughput comparison function which
+// uses reflection to determine if the results match. Probably not good for high RPS endpoints
+// as it'll cause a significant amount of log noise. Can cause panics due to the use of the
+// cmp library (which panics for a number of conditions) so *not* for use without a recover.
+//
+// This function only compares structs/pointers to structs and errors. It will not attempt to compare
+// strings or primitive types and will just state them as not matching.
+func defaultComparisonFn[T any](
+	log log.Logger,
+	scope metrics.Scope,
+	activeRes T,
+	activeErr error,
+	backgroundRes T,
+	backgroundErr error) (isEqual bool) {
+
+	if activeErr != nil && backgroundErr != nil {
+		errDiff := cmp.Diff(activeErr, backgroundErr, cmpopts.EquateErrors())
+		if errDiff != "" {
+			log.Warn("mismatched result with both flows returning an error", tag.Dynamic("comparison-result-errdiff", errDiff))
+			scope.IncCounter(metrics.ShadowerComparisonDifferenceCount)
+			return false
+		}
+		// else they actually match in both having the same error
+		scope.IncCounter(metrics.ShadowerComparisonMatchCount)
```

Comment:
this needs a bunch more dimensions, but I didn't want to gold-plate it too much upfront, I'm not sure I'll be using this out of the box.

Reply:
generally agreed, though I think it might be worth metric-ing err vs value.  otherwise it's hard to see if a change means one errors more than the other or something.

So like `.Tag({"active": "err" or "value", "background": "err" or "value"})`?  Then "errors match" is "match" + both err, and "one failed" is "different" plus you can see which one errored, etc.